data/preprocessor.js

```diff
@@ -0,0 +1,72 @@
+/* eslint-disable no-use-before-define */
+
+const Preprocessor = (jsonRaw = {}) => {
+  const json = { ...jsonRaw }; // make a copy so we don't corrupt the input
+  const timestampNeeded = checkPropertiesExist({
+    input: json,
+    properties: [
+      'metadata.firstPublished',
+      'metadata.lastUpdated',
+      'content.model.blocks',
+    ],
+  });
+  if (timestampNeeded) {
+    const timestampBlock = {
+      type: 'timestamp',
+      model: {
+        published: json.metadata.firstPublished,
+        updated: json.metadata.lastUpdated,
+      },
+    };
+    const { headlineBlocks, mainBlocks } = splitBlocksByHeadline(
+      json.content.model.blocks,
+    );
+    if (headlineBlocks.length > 0) {
+      // insert timestamp block immediately after headline
+      json.content.model.blocks = [
+        ...headlineBlocks,
+        timestampBlock,
+        ...mainBlocks,
+      ];
+    } else {
+      // put timestamp block in as the first element
+      json.content.model.blocks.unshift(timestampBlock);
+    }
+  }
+  return json;
+};
+
+export const checkPropertiesExist = ({ input, properties }) =>
+  properties.reduce(
+    (validates, prop) => validates && validate(input, prop),
+    true,
+  );
+
+const validate = (input, prop) => {
```

Comment:
Whats this trying to do?

Reply:
Have refactored, hopefully it now makes sense.